File: javascript/ambar.js
// ===============================
// Custom Exception
// ===============================
class KuantumCokusuException extends Error {
    constructor(id) {
        super(`KUANTUM ÇÖKÜŞÜ! Nesne patladı: ${id}`);
        this.name = "KuantumCokusuException";
        this.nesneId = id;
    }
}

// ===============================
// Abstract Class (KuantumNesnesi)
// ===============================
class KuantumNesnesi {
    constructor(id, stabilite, tehlikeSeviyesi) {
        // Soyut sınıf kontrolü
        if (new.target === KuantumNesnesi) {
            throw new Error("KuantumNesnesi soyut sınıftır, direkt örnek oluşturulamaz!");
        }
        this._id = id;
        this._stabilite = 0;
        this._tehlikeSeviyesi = 0;
        
        // Setter'lar üzerinden atama (encapsulation)
        this.stabilite = stabilite;
        this.tehlikeSeviyesi = tehlikeSeviyesi;
    }

    // ID getter (sadece okunabilir)
    get id() {
        return this._id;
    }

    // Stabilite getter/setter (Encapsulation: 0-100 arası kısıtlama)
    get stabilite() {
        return this._stabilite;
    }

    set stabilite(value) {
        if (typeof value !== 'number' || isNaN(value)) {
            throw new Error("Stabilite sayısal değer olmalıdır!");
        }
        // 0-100 arası kısıtlama
        if (value < 0) {
            this._stabilite = 0;
        } else if (value > 100) {
            this._stabilite = 100;
        } else {
            this._stabilite = value;
        }
    }

    // TehlikeSeviyesi getter/setter (Encapsulation: 1-10 arası)
    get tehlikeSeviyesi() {
        return this._tehlikeSeviyesi;
    }

    set tehlikeSeviyesi(value) {
        if (typeof value !== 'number' || value < 1 || value > 10) {
            throw new Error("TehlikeSeviyesi 1-10 arası olmalıdır!");
        }
        this._tehlikeSeviyesi = Math.floor(value);
    }

    // Abstract metot (override edilmeli)
    analizEt() {
        throw new Error("analizEt() metodu alt sınıflarda override edilmelidir!");
    }

    // Durum bilgisi
    durumBilgisi() {
        return `ID: ${this.id} | Stabilite: ${this.stabilite.toFixed(1)} | Tehlike: ${this.tehlikeSeviyesi} | Tip: ${this.constructor.name}`;
    }

    // Stabilite kontrolü (her analiz sonrası)
    _stabiliteKontrol() {
        if (this.stabilite <= 0) {
            throw new KuantumCokusuException(this.id);
        }
    }
}

// ===============================
// Interface (IKritik) - Mixin Pattern
// ===============================
const IKritik = Base => class extends Base {
    acilDurumSogutmasi() {
        const eskiStabilite = this.stabilite;
        this.stabilite += 50; // Setter otomatik max 100 yapacak
        console.log(`[SOĞUTMA] ${this.id}: ${eskiStabilite.toFixed(1)} -> ${this.stabilite.toFixed(1)}`);
    }
    
    // IKritik interface'ini implement ettiğini belirtmek için
    get isKritik() {
        return true;
    }
};

// ===============================
// Concrete Classes (Kalıtım & Polimorfizm)
// ===============================

// VeriPaketi - IKritik DEĞİL
class VeriPaketi extends KuantumNesnesi {
    analizEt() {
        console.log("Veri içeriği okundu.");
        this.stabilite -= 5;
        this._stabiliteKontrol();
    }
    
    get isKritik() {
        return false;
    }
}

// KaranlikMadde - IKritik
class KaranlikMadde extends IKritik(KuantumNesnesi) {
    analizEt() {
        console.log("Karanlık madde analiz ediliyor...");
        this.stabilite -= 15;
        this._stabiliteKontrol();
    }
}

// AntiMadde - IKritik (En tehlikeli)
class AntiMadde extends IKritik(KuantumNesnesi) {
    analizEt() {
        console.log("EVRENİN DOKUSU TİTRİYOR...");
        this.stabilite -= 25;
        this._stabiliteKontrol();
    }
}

// ===============================
// Utility Functions
// ===============================
function rastgeleID() {
    return "QN-" + Math.random().toString(36).substring(2, 8).toUpperCase();
}

function rastgeleNesneUret() {
    const tipler = [VeriPaketi, KaranlikMadde, AntiMadde];
    const TipClass = tipler[Math.floor(Math.random() * tipler.length)];
    
    const stabilite = Math.floor(Math.random() * 40) + 60; // 60-100 arası
    const tehlike = Math.floor(Math.random() * 10) + 1;    // 1-10 arası
    
    return new TipClass(rastgeleID(), stabilite, tehlike);
}

// ===============================
// Main Application
// ===============================
const readline = require("readline");
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

let envanter = [];

function menu() {
    console.log(`
${"=".repeat(50)}
   KUANTUM AMBARI KONTROL PANELİ  
${"=".repeat(50)}

1. Yeni Nesne Ekle (Rastgele)
2. Tüm Envanteri Listele
3. Nesneyi Analiz Et (ID)
4. Acil Durum Soğutması Yap (IKritik)
5. Çıkış

Envanter: ${envanter.length} nesne
${"=".repeat(50)}
`);
}

// Her işlemi try-catch ile sarmalama fonksiyonu
function guvenliIslem(callback) {
    try {
        callback();
    } catch (e) {
        if (e instanceof KuantumCokusuException) {
            console.log("\n" + "!".repeat(50));
            console.log("SİSTEM ÇÖKTÜ! TAHLİYE BAŞLATILIYOR...");
            console.log("!" + e.message);
            console.log("!".repeat(50) + "\n");
            rl.close();
            process.exit(1);
        } else {
            console.error("Hata:", e.message);
        }
    }
}

function mainLoop() {
    menu();
    rl.question("Seçiminiz: ", secim => {
        secim = secim.trim();
        
        switch (secim) {
            case "1":
                guvenliIslem(() => {
                    const nesne = rastgeleNesneUret();
                    envanter.push(nesne);
                    const simge = nesne.isKritik ? "False" : "True";
                    console.log(`\n${simge} Yeni nesne eklendi:`);
                    console.log(`   ${nesne.durumBilgisi()}\n`);
                });
                mainLoop();
                break;

            case "2":
                if (envanter.length === 0) {
                    console.log("\nEnvanter boş.\n");
                } else {
                    console.log("\nENVANTER RAPORU:");
                    console.log("-".repeat(80));
                    envanter.forEach((n, idx) => {
                        const kritikBadge = n.isKritik ? " [KRİTİK]" : "";
                        const stabiliteBar = "".repeat(Math.floor(n.stabilite / 10)) + 
                                           "".repeat(10 - Math.floor(n.stabilite / 10));
                        console.log(`${idx + 1}. ${n.durumBilgisi()}${kritikBadge}`);
                        console.log(`   Stabilite: [${stabiliteBar}]`);
                    });
                    console.log("-".repeat(80) + "\n");
                }
                mainLoop();
                break;

            case "3":
                rl.question("Analiz edilecek ID: ", id => {
                    id = id.trim();
                    guvenliIslem(() => {
                        const nesne = envanter.find(x => x.id === id);
                        if (!nesne) {
                            console.log(`\nNesne bulunamadı: ${id}\n`);
                        } else {
                            console.log(`\nAnaliz başlatılıyor: ${id}`);
                            nesne.analizEt(); // Burada KuantumCokusuException fırlatılabilir
                            console.log(`   ${nesne.durumBilgisi()}`);
                            
                            if (nesne.stabilite < 30) {
                                console.log(`UYARI: Kritik stabilite seviyesi!\n`);
                            } else {
                                console.log();
                            }
                        }
                    });
                    mainLoop();
                });
                return;

            case "4":
                rl.question("Soğutulacak ID: ", id => {
                    id = id.trim();
                    guvenliIslem(() => {
                        const nesne = envanter.find(x => x.id === id);
                        if (!nesne) {
                            console.log(`\nNesne bulunamadı: ${id}\n`);
                        } else if (nesne.isKritik && typeof nesne.acilDurumSogutmasi === "function") {
                            console.log(`\nAcil soğutma uygulanıyor: ${id}`);
                            nesne.acilDurumSogutmasi();
                            console.log(`   ${nesne.durumBilgisi()}\n`);
                        } else {
                            console.log(`\nBu nesne soğutulamaz! (IKritik değil)\n`);
                        }
                    });
                    mainLoop();
                });
                return;

            case "5":
                console.log("\nGüvenli çıkış yapılıyor...\n");
                rl.close();
                return;

            default:
                console.log("\nGeçersiz seçim!\n");
                mainLoop();
        }
    });
}

// Program başlangıcı
console.log("\nKuantum Ambarı sistemi başlatılıyor...\n");
mainLoop();

File: python/ambar.py
from __future__ import annotations
from abc import ABC, abstractmethod
from typing import List
import random
import sys

class KuantumCokusuException(Exception):
    def __init__(self, nesne_id: str):
        super().__init__(f"Kuantum Çöküşü! Patlayan nesne ID: {nesne_id}")
        self.nesne_id = nesne_id

class IKritik(ABC):
    @abstractmethod
    def acil_durum_sogutmasi(self) -> None:
        pass

class KuantumNesnesi(ABC):
    def __init__(self, id_: str, stabilite: float, tehlike_seviyesi: int):
        self._id = id_
        self._stabilite = 100.0  # initialize before setter to avoid attribute errors
        self._tehlike_seviyesi = 1
        self.id = id_
        self.stabilite = stabilite
        self.tehlike_seviyesi = tehlike_seviyesi

    @property
    def id(self) -> str:
        return self._id

    @id.setter
    def id(self, value: str) -> None:
        if not value or not isinstance(value, str):
            raise ValueError("ID geçersiz.")
        self._id = value

    @property
    def stabilite(self) -> float:
        return self._stabilite

    @stabilite.setter
    def stabilite(self, value: float) -> None:
        # Stabilite her ayarlamada 0-100 arasında tutulur (clamp).
        if value is None:
            raise ValueError("Stabilite boş olamaz.")
        if value > 100:
            self._stabilite = 100.0
        elif value < 0:
            self._stabilite = 0.0
        else:
            self._stabilite = float(value)

    @property
    def tehlike_seviyesi(self) -> int:
        return self._tehlike_seviyesi

    @tehlike_seviyesi.setter
    def tehlike_seviyesi(self, value: int) -> None:
        if not isinstance(value, int) or value < 1 or value > 10:
            raise ValueError("TehlikeSeviyesi 1-10 arası olmalıdır.")
        self._tehlike_seviyesi = value

    @abstractmethod
    def analiz_et(self) -> None:
        pass

    def durum_bilgisi(self) -> str:
        return f"ID: {self.id}, Stabilite: {self.stabilite:.1f}, Tehlike: {self.tehlike_seviyesi}, Tip: {self.__class__.__name__}"

    # Yardımcı: stabilite düşüşleri sonrası çöküş kontrolü
    def _kuantum_cokusu_kontrol(self) -> None:
        if self.stabilite <= 0:
            raise KuantumCokusuException(self.id)

class VeriPaketi(KuantumNesnesi):
    # IKritik değildir.
    def analiz_et(self) -> None:
        # Stabilite 5 düşer, mesaj verir
        print("Veri içeriği okundu.")
        self.stabilite -= 5
        self._kuantum_cokusu_kontrol()

class KaranlikMadde(KuantumNesnesi, IKritik):
    def analiz_et(self) -> None:
        self.stabilite -= 15
        self._kuantum_cokusu_kontrol()

    def acil_durum_sogutmasi(self) -> None:
        self.stabilite = min(100.0, self.stabilite + 50)

class AntiMadde(KuantumNesnesi, IKritik):
    # Çok tehlikelidir. Her işlemde 25 düşer.
    def analiz_et(self) -> None:
        print("Evrenin dokusu titriyor...")
        self.stabilite -= 25
        self._kuantum_cokusu_kontrol()

    def acil_durum_sogutmasi(self) -> None:
        self.stabilite = min(100.0, self.stabilite + 50)

def rastgele_nesne_uret(sayac: int) -> KuantumNesnesi:
    tip = random.choice(["veri", "karanlik", "anti"])
    id_ = f"N{sayac:04d}"
    stabilite = random.uniform(50, 100)  # başlangıçta makul stabilite
    tehlike = random.randint(1, 10)
    if tip == "veri":
        return VeriPaketi(id_, stabilite, tehlike)
    elif tip == "karanlik":
        return KaranlikMadde(id_, stabilite, tehlike)
    else:
        return AntiMadde(id_, stabilite, tehlike)

def nesneyi_bul(envanter: List[KuantumNesnesi], id_: str) -> KuantumNesnesi | None:
    for n in envanter:
        if n.id == id_:
            return n
    return None

def main():
    envanter: List[KuantumNesnesi] = []
    sayac = 1
    while True:
        try:
            print("\nKUANTUM AMBARI KONTROL PANELİ")
            print("1. Yeni Nesne Ekle (Rastgele)")
            print("2. Tüm Envanteri Listele")
            print("3. Nesneyi Analiz Et (ID)")
            print("4. Acil Durum Soğutması Yap (ID, sadece IKritik)")
            print("5. Çıkış")
            secim = input("Seçiminiz: ").strip()

            if secim == "1":
                nesne = rastgele_nesne_uret(sayac)
                sayac += 1
                envanter.append(nesne)
                print(f"Eklendi -> {nesne.durum_bilgisi()}")

            elif secim == "2":
                if not envanter:
                    print("Envanter boş.")
                else:
                    for n in envanter:
                        print(n.durum_bilgisi())

            elif secim == "3":
                id_ = input("Analiz edilecek ID: ").strip()
                n = nesneyi_bul(envanter, id_)
                if not n:
                    print("Nesne bulunamadı.")
                else:
                    n.analiz_et()
                    print("Analiz tamam. " + n.durum_bilgisi())

            elif secim == "4":
                id_ = input("Soğutma uygulanacak ID: ").strip()
                n = nesneyi_bul(envanter, id_)
                if not n:
                    print("Nesne bulunamadı.")
                else:
                    if isinstance(n, IKritik):
                        n.acil_durum_sogutmasi()
                        print("Soğutma uygulandı. " + n.durum_bilgisi())
                    else:
                        print("Bu nesne soğutulamaz!")

            elif secim == "5":
                print("Güvenli çıkış yapılıyor...")
                break

            else:
                print("Geçersiz seçim.")

        except KuantumCokusuException as ex:
            print("SİSTEM ÇÖKTÜ! TAHLİYE BAŞLATILIYOR...")
            print(str(ex))
            sys.exit(1)
        except Exception as ex:
            # Beklenmeyen hatalar için nazik mesaj
            print(f"Hata: {ex}")

if __name__ == "__main__":
    main()
